### ISSUE TITLE
Fix stateStorageStrategy localStorage not increasing session counters

### ISSUE THREAD

**Issue by christoph-buente** [snowplow/snowplow-javascript-tracker#718]
We are looking into changing the `stateStorageStrategy` from `cookie` to `localstorage`. The current mechanism to detect a new session is the absence of the _sp_ses cookie which expires after the configured `sessionCookieTimeout`. The problem is, that localstorage does not have an expiry mechanism, so the the _sp_ses.xxxx key in local storage stays forever.

The timestamps are properly written to the _sp_id.xxxx key in local storage. And the timestamps are actually enough to detect a beginning of a new session. If current timestamp and last visit time stamp are longer apart from each other than the session duration, the counter could be incremented.

And then we could get rid of _sp_ses cookie/localstorage key.

What do you think?

**Comment by christoph-buente** [comment]
@mhadam any thoughts?

**Comment by mhadam** [comment]
Hi Christophe! That seems reasonable, and no problems really occur to me at first glance. I’ll try to get this in a new release within the next couple weeks.

**Comment by christoph-buente** [comment]
Thank you!

**Comment by grzegorzewald** [comment]
The issue has been addressed at Snowplow forums (https://discourse.snowplowanalytics.com/t/session-not-tracked-correctly/3026). Can it be prioritized? Especially as it seems to be slight workaround for ITP2.2.

**Comment by goodits** [comment]
Hi @grzegorzewald we can take a look at this next week and get back to you!

**Comment by christoph-buente** [comment]
Thanks Steve, much appreciated

**Comment by goodits** [comment]
I owe everyone on this thread an update! Sincere apologies for the radio silence.

One thing that has held things up on this is that we ideally don't want to ship a change before we've completed a new suite of integration tests for the tracker. We've had to roll patches to a few tracker releases recently following up on regression bugs, and even a few is a few too many! With histories and configurations as long as our trackers, they're hard to manually test. So we're investing time in automated tests that build confidence that the release is going to work, end to end, as intended and in perpetuity. We'd like to get at least the beginnings of these tests and cover off some critical paths before rolling out any change to the storage strategy. 

We're in a position to prioritise both the tests and this change now, so we'll hopefully not keep you waiting too long. I've made a note to keep this thread updated with progress. Thank you for your patience!
